File: Server/src/DTOs/Location.java
package DTOs;

import java.util.Objects;

public class Location {
    private int location_id;
    private String address;
    private double latitude;
    private double longitude;

    public Location(int location_id, String address, double latitude, double longitude)
    {
        this.location_id = location_id;
        this.address = address;
        this.latitude = latitude;
        this.longitude = longitude;
    }
    
    public Location(String address, double latitude, double longitude)
    {
        this.address = address;
        this.latitude = latitude;
        this.longitude = longitude;
    }
    
    public Location()
    {
        this.location_id = 0;
        this.address = "N/A";
        this.latitude = 0;
        this.longitude = 0;
    }
    
    public int getLocation_id() {
        return location_id;
    }

    public void setLocation_id(int location_id) {
        this.location_id = location_id;
    }

    public String getAddress() {
        return address;
    }

    public void setAddress(String address) {
        this.address = address;
    }

    public double getLatitude() {
        return latitude;
    }

    public void setLatitude(double latitude) {
        this.latitude = latitude;
    }

    public double getLongitude() {
        return longitude;
    }

    public void setLongitude(double longitude) {
        this.longitude = longitude;
    }

    @Override
    public int hashCode() {
        int hash = 3;
        hash = 67 * hash + this.location_id;
        hash = 67 * hash + Objects.hashCode(this.address);
        hash = 67 * hash + Objects.hashCode(this.latitude);
        hash = 67 * hash + Objects.hashCode(this.longitude);
        return hash;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
        {
            return true;
        }
        if (obj == null)
        {
            return false;
        }
        if (getClass() != obj.getClass())
        {
            return false;
        }
        final Location other = (Location) obj;
        if (this.location_id != other.location_id)
        {
            return false;
        }
        if (!Objects.equals(this.address, other.address))
        {
            return false;
        }
        if (!Objects.equals(this.latitude, other.latitude))
        {
            return false;
        }
        if (!Objects.equals(this.longitude, other.longitude))
        {
            return false;
        }
        return true;
    }

    @Override
    public String toString() {
        return "Location{" + "location_id=" + location_id + ", address=" + address + ", latitude=" + latitude + ", longitude=" + longitude + '}';
    }
    
    
}

File: Client/models/Location.php
<?php require_once('Model.php') ?>
<?php

class Location
{
    private $location_id;
    private $address;
    private $latitude;
    private $longitude;

    public function __construct($args)
    {
        if (!is_array($args)) {
            throw new Exception('Location constructor requires an array');
        }
        $this->location_id   = $args['location_id']      ?? NULL;
        $this->address       = $args['address']          ?? NULL;
        $this->latitude      = $args['latitude']         ?? NULL;
        $this->longitude     = $args['longitude']        ?? NULL;
    }


    public function getLocation_id()
    {
        return $this->location_id;
    }
    public function getAddress()
    {
        return $this->address;
    }
    public function getLatitude()
    {
        return $this->latitude;
    }
    public function getLongitude()
    {
        return $this->longitude;
    }

    public function setAddress($address)
    {
        if ($address === NULL) {
            $this->address = $this->address;
            return;
        }
        $this->address = $address;
    }

    
    public static function returnLatLongById($db, $location_id)
    {
        $statement = $db->prepare("select latitude, longitude from location where location_id = :location_id");
        $statement->execute([
            'location_id' => $location_id
        ]);
        $id = $statement->fetch();
        $statement->closeCursor();
        return $id;
    }

    public static function returnLocation_idByAddress($db, $address)
    {
        $statement = $db->prepare("select location_id from location where address = :address");
        $statement->execute([
            'address' => $address
        ]);
        $id = $statement->fetch();
        $statement->closeCursor();
        return $id;
    }

    public static function returnAddressById($db, $location_id)
    {
        $statement = $db->prepare("select * from location where location_id = :location_id");
        $statement->execute([
            'location_id' => $location_id
        ]);
        $address = $statement->fetch();
        $statement->closeCursor();
        return $address;
    }
    public static function calculateDistance($db, $location_id, $currentUserLat, $currentUserLong)
    {
        $statement = $db->prepare("select calculate_distance((select latitude from location where location_id = :location_id), (select longitude from location where location_id = :location_id), :latitude, :longitude, 'K')");
        $statement->execute([
            'location_id' => $location_id,
            'latitude' => $currentUserLat,
            'longitude' => $currentUserLong
        ]);
        $distance = $statement->fetch();
        $statement->closeCursor();
        return $distance;
    }

    public static function addLocation($db, $location)
    {
        $statement = $db->prepare("insert into location(address, latitude, longitude) values (:address, :latitude, :longitude); ");
        $statement->execute([
            'address'   => $location->getAddress(),
            'latitude'  => $location->getLatitude(),
            'longitude' => $location->getLongitude()
        ]);
        $statement->closeCursor();
    }


    public static function updateLocation($db, $location)
    {
        $statement = $db->prepare("update location set address = :address, latitude = :latitude, longitude = :longitude");
        $statement->execute([
            'address'   => $location->getAddress(),
            'latitude'  => $location->getLatitude(),
            'longitude' => $location->getLongitude()
        ]);
        $statement->closeCursor();
    }

    public static function deleteLocation($db, $id)
    {
        $statement = $db->prepare("delete from location where location_id = $id");
        $statement->execute();
        $statement->closeCursor();
    }
}


?>